plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,337 @@
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+
+use crate::hash::hash_types::{RichField, NUM_HASH_OUT_ELTS};
+use crate::hash::merkle_proofs::MerkleProof;
+use crate::hash::merkle_tree::{
+    capacity_up_to_mut, fill_digests_buf, merkle_tree_prove, MerkleCap,
+};
+use crate::plonk::config::{GenericHashOut, Hasher};
+use crate::util::log2_strict;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
+    /// The data stored in the Merkle tree leaves.
+    pub leaves: Vec<Vec<Vec<F>>>,
+
+    /// Merkle tree node hashes, analogous to `digests` in `MerkleTree`.
+    pub digests: Vec<H::Hash>,
+
+    /// Represents the roots of the Merkle tree. This allows for using any layer as the root of the tree.
+    pub cap: MerkleCap<F, H>,
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
+    /// Each element in the `leaves` vector represents a matrix (a vector of vectors).
+    /// The height of each matrix should be a power of two.
+    /// The `leaves` vector should be sorted by matrix height, from tallest to shortest, with no duplicate heights.
+    // TODO: FieldMerkleTree does not handle duplicates; this is deferred to the caller.  Revisit when implementing batch FRI to potentially optimize.
+    pub fn new(mut leaves: Vec<Vec<Vec<F>>>, cap_height: usize) -> Self {
+        assert!(!leaves.is_empty());
+        assert!(leaves.iter().all(|leaf| leaf.len().is_power_of_two()));
+        assert!(leaves
+            .windows(2)
+            .all(|pair| { pair[0].len() > pair[1].len() }));
+
+        let leaves_len = leaves[0].len();
+        let log2_leaves_len = log2_strict(leaves_len);
+        assert!(
```

Comment:
I know it's not in the original code either, but maybe assert that the last log2 is >= cap _height?

Reply:
Yes! We should compare last log2.